NeoML/test/src/ReferenceDnnTest.cpp

```diff
@@ -0,0 +1,232 @@
+	}
+
+	IPerformanceCounters* counters(MathEngine().CreatePerformanceCounters());
+	counters->Synchronise();
+
+	getTestDnns(dnns, randoms, useReference, numOfThreads);
+	runMultithreadInference(dnns, numOfThreads);
+
+	counters->Synchronise();
+	std::cerr
+		<< '\n' << "Time: " << (double((*counters)[0].Value) / 1000000) << " ms."
+		<< '\t' << "Peak.Mem: " << (double(MathEngine().GetPeakMemoryUsage()) / 1024 / 1024) << " MB \n";
+
+
+	delete counters;
+	// delete references first
+	for(int i = 1; i < numOfThreads; ++i) {
+		delete dnns[i];
+	}
+	delete dnns[0];
+}
+
+} // namespace NeoMLTest
+
+//------------------------------------------------------------------------------------------------
+
+TEST(ReferenceDnnTest, ReferenceDnnInferenceTest)
+{
+	const int numOfThreads = 4;
+	CArray<CDnn*> dnns;
+	CArray<CRandom> randoms;
+
+	for(int i = 0; i < numOfThreads; ++i) {
+		randoms.Add(CRandom(i));
+	}
+
+	getTestDnns(dnns, randoms, true, numOfThreads);
+	runMultithreadInference(dnns, numOfThreads);
+
+	for(int i = 0; i < numOfThreads - 1; ++i) {
+		EXPECT_TRUE(CompareBlobs(
+			*(static_cast<CSourceLayer*>(dnns[i]->GetLayer("in").Ptr())->GetBlob()),
+			*(static_cast<CSourceLayer*>(dnns[i + 1]->GetLayer("in").Ptr())->GetBlob()))
+		);
+
+		EXPECT_TRUE(CompareBlobs(
+			*(static_cast<CSinkLayer*>(dnns[i]->GetLayer("sink").Ptr())->GetBlob()),
+			*(static_cast<CSinkLayer*>(dnns[i + 1]->GetLayer("sink").Ptr())->GetBlob()))
+		);
+	}
+
+	// delete references first
+	for(int i = 1; i < numOfThreads; ++i) {
+		delete dnns[i];
+	}
+	delete dnns[0];
+}
+
+TEST(ReferenceDnnTest, CDnnReferenceRegisterTest)
+{
+	// Implement scenario - learn dnn, use multihtread inference, learn again
+	const int numOfThreads = 4;
+
+	CArray<CReferenceDnnTestParam> taskParams;
+	CArray<CRandom> randoms;
+	CArray<CDnn*> dnns;
+	CRandom randomInit(0);
+
+	// 1.Create and learn dnn
+	randoms.Add(CRandom(0));
+	dnns.Add(createDnn(randoms[0]));
+
+	CPtr<CDnnBlob> srcBlob = CDnnBlob::CreateTensor(MathEngine(), CT_Float, { 1, 1, 1, 8, 20, 30, 10 });
+	initializeBlob(srcBlob.Ptr(), randomInit, 0, 1);
+	static_cast<CSourceLayer*>(dnns[0]->GetLayer("in").Ptr())->SetBlob(srcBlob);
+
+	CPtr<CSourceLayer> labels = Source(*dnns[0], "labels");
+	CPtr<CDnnBlob> labelBlob = CDnnBlob::CreateTensor(MathEngine(), CT_Float, { 1, 1, 1, 1, 1, 1, 10 });
+	initializeBlob(labelBlob.Ptr(), randomInit, 0, 1);
+	labels->SetBlob(labelBlob);
+
+	CPtr<CL1LossLayer> loss = L1Loss()("loss", dnns[0]->GetLayer("fc3").Ptr(), labels.Ptr());
+
+	CPtr<CDnnSimpleGradientSolver> solver = new CDnnSimpleGradientSolver(MathEngine());
+	solver->SetLearningRate(1e-5f);
+	solver->SetMomentDecayRate(0.f);
+	dnns[0]->SetSolver(solver);
+
+	for(int i = 0; i < 10; ++i) {
+		dnns[0]->RunAndLearnOnce();
+	}
+
+	// 2. Run mulithread inference
+	dnns[0]->DeleteLayer("labels");
+	dnns[0]->DeleteLayer("loss");
+
+	for(int i = 1; i < numOfThreads; ++i) {
+		randoms.Add(CRandom(i));
+		dnns.Add(dnns[0]->CreateReferenceDnn());
```

Comment:
Should to add a test to check user random works and gets different output for reference and original nets?
I have a suggestion that random do not interfere on inference...

Reply:
Yep, probably in future ( for gans for example) but not in the tests definitely :)
The only think i came up with is to use dnns' random generators to fill some blobs and compare them

Reply:
No need such a test